pkg/build/sbom.go

```diff
@@ -0,0 +1,180 @@
+// Copyright 2022 Chainguard, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package build
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+
+	osr "github.com/dominodatalab/os-release"
+	"github.com/pkg/errors"
+	"gitlab.alpinelinux.org/alpine/go/pkg/repository"
+)
+
+// TODO(kaniini): Move most of this over to gitlab.alpinelinux.org/alpine/go.
+type document struct {
+	BOMFormat    string       `json:"bomFormat"`
+	SpecVersion  string       `json:"specVersion"`
+	Version      int          `json:"version"`
+	Components   []component  `json:"components,omitempty"`
+	Dependencies []dependency `json:"dependencies,omitempty"`
+}
+
+type component struct {
+	BOMRef             string              `json:"bom-ref"`
+	Type               string              `json:"type"`
+	Name               string              `json:"name"`
+	Version            string              `json:"version"`
+	Description        string              `json:"description"`
+	PUrl               string              `json:"purl"`
+	ExternalReferences []externalReference `json:"externalReferences,omitempty"`
+	Licenses           []license           `json:"licenses,omitempty"`
+	Components         []component         `json:"components,omitempty"`
+}
+
+type license struct {
+	Expression string `json:"expression"`
+}
+
+type externalReference struct {
+	URL  string `json:"url"`
+	Type string `json:"type"`
+}
+
+type dependency struct {
+	Ref       string   `json:"ref"`
+	DependsOn []string `json:"dependsOn"`
+}
+
+func bomRef(ns string, pkg *repository.Package) string {
+	return fmt.Sprintf("pkg:apk/%s/%s", ns, pkg.Name)
+}
+
+func bomPurl(ns string, pkg *repository.Package) string {
+	return fmt.Sprintf("pkg:apk/%s/%s@%s", ns, pkg.Name, pkg.Version)
+}
+
+func (bc *Context) GenerateSBOM() error {
+	log.Printf("generating SBOM")
+
+	installedDB, err := os.Open(filepath.Join(bc.WorkDir, "lib", "apk", "db", "installed"))
```

Comment:
this is leaking, it needs a call to `defer installedDB.Close()` after the err check

Reply:
It's not: https://gitlab.alpinelinux.org/alpine/go/-/blob/master/pkg/repository/apkindex.go#L22

Reply:
oh, I missed that! Thanks :bow: